MATLAB/update_FC_state.m

```diff
@@ -12,91 +12,149 @@
 %   update state with commanded wheel torque
 
 global const
-state=initial_state;
-actuators= struct();
 
-%% Attitude determination
+%% Attitude Determination %%
 
-sat2sun_eci= env_sun_vector(sensor_readings.time);
-mag_ecef= env_magnetic_field(sensor_readings.time,sensor_readings.position_ecef);
-[quat_ecef_eci,~]=env_earth_attitude(sensor_readings.time);
-quat_eci_ecef= utl_quat_conj(quat_ecef_eci);
-mag_eci =utl_rotateframe(quat_eci_ecef,mag_ecef);
-quat_body_eci=utl_triad(sat2sun_eci,mag_eci,sensor_readings.sat2sun_body,sensor_readings.magnetometer_body);
-angular_momentum_body= const.JB*sensor_readings.gyro_body+sensor_readings.wheel_momentum_body;
+estimator.sat2sun_eci= env_sun_vector(sensor_readings.time);
+estimator.mag_ecef= env_magnetic_field(sensor_readings.time,sensor_readings.position_ecef);
+[estimator.quat_ecef_eci,~]=env_earth_attitude(sensor_readings.time);
+estimator.quat_eci_ecef= utl_quat_conj(estimator.quat_ecef_eci);
+estimator.mag_eci =utl_rotateframe(estimator.quat_eci_ecef,estimator.mag_ecef);
+estimator.sat2sun_body= sensor_readings.sat2sun_body;
+estimator.quat_body_eci=utl_triad(estimator.sat2sun_eci,estimator.mag_eci,sensor_readings.sat2sun_body,sensor_readings.magnetometer_body);
+estimator.angular_rate_body= sensor_readings.gyro_body;
+estimator.angular_momentum_body= const.JB*estimator.angular_rate_body+sensor_readings.wheel_momentum_body;
+estimator.magnetometer_body= sensor_readings.magnetometer_body;
+estimator.magnetic_field_body=estimator.magnetometer_body;
+estimator.time= sensor_readings.time;
+estimator.position_eci= utl_rotateframe(estimator.quat_eci_ecef,sensor_readings.position_ecef);
+estimator.velocity_eci= utl_rotateframe(estimator.quat_eci_ecef,sensor_readings.velocity_ecef+cross(const.earth_rate_ecef,sensor_readings.position_ecef));
+estimator.target_position_eci=utl_rotateframe(estimator.quat_eci_ecef,sensor_readings.target_position_ecef);
+estimator.target_velocity_eci=utl_rotateframe(estimator.quat_eci_ecef,sensor_readings.target_velocity_ecef+cross(const.earth_rate_ecef,sensor_readings.target_position_ecef));
+estimator.tumbling= ~(norm(estimator.angular_momentum_body)<const.MAXWHEELRATE*const.JWHEEL*0.7);
+estimator.low_power= false;
+estimator.eclipse= env_eclipse(estimator.position_eci,estimator.sat2sun_eci);
+estimator.valid_gps=all(isfinite([estimator.position_eci; estimator.velocity_eci; estimator.time]));
+estimator.valid_target_gps=estimator.valid_gps && all(isfinite([estimator.target_position_eci; estimator.target_velocity_eci; estimator.time]));
+
+if estimator.valid_gps && ~estimator.eclipse && ~all(isfinite(estimator.sat2sun_body))
+    estimator.sat2sun_body= [0;0;-1]; % assume the sun is coming from the docking face
+    estimator.quat_body_eci=utl_triad(estimator.sat2sun_eci,estimator.mag_eci,estimator.sat2sun_body,estimator.magnetic_field_body);
+end
```

Comment:
Throwing this in an attitude estimator function and gps estimation function may be a good idea?